# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1528830379 14400
#      Tue Jun 12 15:06:19 2018 -0400
# Branch stable
# Node ID 1c4ffd4f940496e200009deb5fb8111cce40e03c
# Parent  265d3a59da270972cb39a80193816e1c2a920ae3
create valid tmp scope for who -file (bug #54030)

* variables.cc (do_who): Provide name to symbol_scope contstructor to
create valid scope.
(Fwho): New test.

diff --git a/libinterp/corefcn/variables.cc b/libinterp/corefcn/variables.cc
--- a/libinterp/corefcn/variables.cc
+++ b/libinterp/corefcn/variables.cc
@@ -1658,17 +1658,17 @@ do_who (octave::interpreter& interp, int
                    my_name.c_str ());
 
           std::string nm = argv[i + 1];
 
           octave::unwind_protect frame;
 
           // Set up temporary scope.
 
-          octave::symbol_scope tmp_scope;
+          octave::symbol_scope tmp_scope ("$dummy_scope$");
 
           symtab.set_scope (tmp_scope);
 
           cs.push (tmp_scope, 0);
           frame.add_method (cs, &octave::call_stack::pop);
 
           octave::feval ("load", octave_value (nm), 0);
 
@@ -1863,16 +1863,34 @@ matching the given patterns.
 {
   int argc = args.length () + 1;
 
   string_vector argv = args.make_argv ("who");
 
   return do_who (interp, argc, argv, nargout == 1);
 }
 
+/*
+%!test
+%! avar = magic (4);
+%! ftmp = [tempname() ".mat"];
+%! unwind_protect
+%!   save (ftmp, "avar");
+%!   vars = whos ("-file", ftmp);
+%!   assert (numel (vars), 1);
+%!   assert (isstruct (vars));
+%!   assert (vars.name, "avar");
+%!   assert (vars.size, [4, 4]);
+%!   assert (vars.class, "double");
+%!   assert (vars.bytes, 128);
+%! unwind_protect_cleanup
+%!   unlink (ftmp);
+%! end_unwind_protect
+*/
+
 DEFMETHOD (whos, interp, args, nargout,
            doc: /* -*- texinfo -*-
 @deftypefn  {} {} whos
 @deftypefnx {} {} whos pattern @dots{}
 @deftypefnx {} {} whos option pattern @dots{}
 @deftypefnx {} {S =} whos ("pattern", @dots{})
 Provide detailed information on currently defined variables matching the
 given patterns.
